src/providers/azure.rs

```diff
@@ -0,0 +1,519 @@
+            "southcentralus" => Region::SouthCentralUs,
+            "northeurope" => Region::NorthEurope,
+            "westeurope" => Region::WestEurope,
+            "eastasia" => Region::EastUsia,
+            "southeastasia" => Region::SouthEastUsia,
+            "japaneast" => Region::JapanEast,
+            "japanwest" => Region::JapanWest,
+            "australiaeast" => Region::AustraliaEast,
+            "australiasoutheast" => Region::AustraliaSoutheast,
+            "australiacentral" => Region::AustraliaCentral,
+            "brazilsouth" => Region::BrazilSouth,
+            "southindia" => Region::SouthIndia,
+            "centralindia" => Region::CentralIndia,
+            "westindia" => Region::WestIndia,
+            "canadacentral" => Region::CanadaCentral,
+            "canadaeast" => Region::CanadaEast,
+            "westus2" => Region::WestUs2,
+            "westcentralus" => Region::WestCentralus,
+            "uksouth" => Region::UkSouth,
+            "ukwest" => Region::UkWest,
+            "koreacentral" => Region::KoreaCentral,
+            "koreasouth" => Region::KoreaSouth,
+            "francecentral" => Region::FranceCentral,
+            "southafricanorth" => Region::SouthAfricaNorth,
+            "uaenorth" => Region::UaeNorth,
+            "germanywestcentral" => Region::GermanyWestCentral,
+            u => bail!("Unknown azure region {}", u),
+        })
+    }
+}
+
+mod azcmd {
+    use super::Error;
+    use super::Region;
+    use failure::ResultExt;
+    use serde::{Deserialize, Serialize};
+    use std::process::Command;
+
+    pub(crate) fn create_resource_group(r: Region, name: &str) -> Result<(), Error> {
+        let out = Command::new("az")
+            .args(&[
+                "group",
+                "create",
+                "--name",
+                name,
+                "--location",
+                &r.to_string(),
+            ])
+            .output()?;
+
+        if !out.status.success() {
+            bail!("Failed to create resource group {} in region {:?}", name, r)
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn create_vm(
+        rg: &str,
+        name: &str,
+        size: &str,
+        image: &str,
+        username: &str,
+    ) -> Result<String, Error> {
+        #[allow(non_snake_case)]
+        #[derive(Debug, Deserialize, Serialize)]
+        struct VmCreateOut {
+            powerState: String,
+            publicIpAddress: String,
+            resourceGroup: String,
+        }
+
+        let out = Command::new("az")
+            .args(&[
+                "vm",
+                "create",
+                "--resource-group",
+                rg,
+                "--name",
+                name,
+                "--image",
+                image,
+                "--size",
+                size,
+                "--admin-username",
+                username,
+                "--ssh-key-value",
+                "@~/.ssh/id_rsa.pub",
```

Comment:
Should this be hardcoded? Can Azure do ephemeral key generation for the user like AWS can?

Reply:
For example, I don't have an RSA key...

Reply:
I don't know how to do this ephemerally for Azure - I think the setup process for Azure involves making a key anyway.

Reply:
Hmm, can we at least make the path configurable then?

Reply:
https://docs.microsoft.com/en-us/azure/virtual-machines/linux/mac-create-ssh-keys#create-an-ssh-key-pair

What we maybe could do is use `--generate-ssh-keys` and `--ssh-dest-key-path` to make a key, then manually delete it in `impl Drop`.

Reply:
https://github.com/akshayknarayan/tsunami/tree/azure-key-gen, but as we discussed offline I can't test it currently.

Reply:
That looks good to me!